### DISCUSSION TITLE
Can I make selection box in fabric js none click sensetive?

### DISCUSSION THREAD

**Discussion by svoboda200786** [fabricjs/fabric.js#7251] [Q&A]
When you have several objects like lines and this objects lay one under another, if you want select the lower one you couldn't do this - will be selected the upper object. Like on this picture. You can't select yellow line - the red one would be selected This is appears because selection area of line if line has angle, for example 45 degrees, is bigger then object and if you click on it, you select this object. But in Adobe Illustrator or Inkscape selectin area is not click sensetive - just if you click directly on line, it would be selected - can I manage the same in fabric?
![444](https://user-images.githubusercontent.com/46110155/127344507-a047f2bf-a6a5-4985-b307-92cb422269f6.png)

**Comment by svoboda200786** [comment]
I think, I found the solution for lines. I use Pythagorean theorem to find all line points and when coordinates of one of them is equal to coordinates of mouse I discard the selected element (it used to be the last added) and set the line as active object
`
	function getBorders(e, line){
	
	    let result = false;
	    let a = line.x1 - line.x2;
	    let b = line.y1 - line.y2; 
	    let a_ = 0, b_ = 0;
	    let delta_a, delta_b;   
	    let length = Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2));
	    delta_a = a / length;
	    delta_b = b / length; 
	    a_ = 0;
	    b_ = 0;      
	    for(let i = 0; i < length; i++){
	        a_ += delta_a;
	        b_ += delta_b;
	        let luft = 3;
	        if((e.pointer.x < Math.round(line.x2 + a_) + luft && e.pointer.x > Math.round(line.x2 + a_) - luft) && (e.pointer.y < Math.round(line.y2 + b_) + luft && e.pointer.y > Math.round(line.y2 + b_) - luft)){
	            result = true;          
        	}                       
	    }
	    return result;
	}
`
	
As all shapes except circles and curves consist of lines we can apply this method to others. And for circles and curves we can do somethig with trigonometry.
	
[Example in codepen. Try to select parts of lines in rectangle with checked and unchecked checkbox](https://codepen.io/svoboda200786/pen/poPLWoW)

Try to select lines by clicking in this red marked area with checkbox and without
![Try to select lines by clicking in this area](https://user-images.githubusercontent.com/46110155/127442392-0a27cfb3-6ee2-46c0-afe2-b012d8ba53bf.jpg)

Video illustration:
https://disk.yandex.ru/d/bu7fulyttBVBfw

**Comment by i-prikot** [comment]
Hi.

Have a look at this canvas property http://fabricjs.com/docs/fabric.Canvas.html#perPixelTargetFind

**Reply by svoboda200786** [reply]
I think it's not the same:-( If PixelTargetFind is true you can select the object by clicking exatly on object, not on boundary box, but you still couldn't select object if it is under other object that was added later, like lines in codepen under the rectangle.

**Reply by svoboda200786** [reply]
I have added this property to my codepen examble above. You can try it yourself. It does not make needed effect. May be I am doing something wrong?

**Reply by ShaMan123** [reply]
As mentioned use `perPixelTargetFind` with [`selectable` property](http://fabricjs.com/docs/fabric.Object.html#selectable)

**Reply by ShaMan123** [reply]
In addition look at `onSelect` to dynamically achieve what you are trying to do:
http://fabricjs.com/docs/fabric.Object.html#onSelect

**Reply by ShaMan123** [reply]
https://codepen.io/shaman123/pen/OJmvQdj

**Reply by svoboda200786** [reply]
> As mentioned use `perPixelTargetFind` with [`selectable` property](http://fabricjs.com/docs/fabric.Object.html#selectable)

You made selectable false? I need objects could be selected by clicking their parts that are covered by other objects. Like parts of this two lines under rectangle.

**Reply by svoboda200786** [reply]
> In addition look at `onSelect` to dynamically achieve what you are trying to do:
> http://fabricjs.com/docs/fabric.Object.html#onSelect

Uncheck checkbox and try to select lines by clicking their parts that are covered by rectangle. The rectangle would be selected.

**Reply by svoboda200786** [reply]
I need both lines and rectangle could be selected. Even I click their parts that are covered by other objects. Intersected parts.

**Reply by svoboda200786** [reply]
![IMG_20210729_091604](https://user-images.githubusercontent.com/46110155/127441309-2f798229-eae7-4fe0-8cfc-c5cf6c5bd267.jpg)
Try to click in this place with unchecked checkbox. Line would not be selected. And I need they to be selected.

**Reply by ShaMan123** [reply]
I don't completely understand the UX.
If I understand correctly you are trying to select an object that is under the currently selected object.
To achieve this you can do a few things:
1. unselect the current object on `mouse:down:before` and or make it unselectable and apply your logic.
2. extend canvas logic for `findTarget` and apply logic there:

> https://codesandbox.io/s/prod-leaf-mjj0p?file=/src/fabric.extend.ts
> `Canvas#findTarget` Logic
_Originally posted by @ShaMan123 in https://github.com/fabricjs/fabric.js/issues/6776#issuecomment-860182752_

**Reply by svoboda200786** [reply]
> I don't completely understand the UX.
> If I understand correctly you are trying to select an object that is under the currently selected object.
> To achieve this you can do a few things:
> 
> 1. unselect the current object on `mouse:down:before` and or make it unselectable and apply your logic.
> 2. extend canvas logic for `findTarget` and apply logic there:
> 
> > https://codesandbox.io/s/prod-leaf-mjj0p?file=/src/fabric.extend.ts
> > `Canvas#findTarget` Logic
> > _Originally posted by @ShaMan123 in [#6776 (comment)](https://github.com/fabricjs/fabric.js/issues/6776#issuecomment-860182752)_

All objects should be selectable. But if you click to the part of object that is covered by another object (like red marked area in second screenshot), the object should be selected too. If you click on one of lines in that area, the would not be selected - will be selected rectangle - It not what I need. In my example rectangle should be selected if you click in those parts of canvas where no lines, In white space I mean.

**Reply by svoboda200786** [reply]
In my codepen I show what is needed. with checkbox - right, without - wrong. But I want solution by fabric.

**Reply by ShaMan123** [reply]
What I suggested might work

**Reply by svoboda200786** [reply]
> 
> 
> What I suggested might work

May be video can clarify my question:
https://disk.yandex.ru/d/bu7fulyttBVBfw

**Reply by i-prikot** [reply]
@svoboda200786, https://codepen.io/i-prikot/pen/MWmVVKR this example is similar to the example from the video

**Reply by ShaMan123** [reply]
@i-prikot did it!
found a typo PixelTargetFind -> perPixelTargetFind
https://codepen.io/shaman123/pen/WNjJwqM

**Reply by svoboda200786** [reply]
> 
> 
> @svoboda200786, https://codepen.io/i-prikot/pen/MWmVVKR this example is similar to the example from the video

Great! It's exactly the same. Thank's a lot:-)

**Reply by svoboda200786** [reply]
> 
> 
> @i-prikot did it!
> found a typo PixelTargetFind -> perPixelTargetFind
> https://codepen.io/shaman123/pen/WNjJwqM

Yes, thank's a lot! Yesterday I added  "perPixelTargetFind: true" only to lines, so it didn't work. Now I see.
